landlab/components/carbonate/carbonate_producer.py

```diff
@@ -0,0 +1,219 @@
+        },
+    }
+
+    def __init__(
+        self,
+        grid,
+        extinction_coefficient=1.0,
+        max_carbonate_production_rate=1000.0,
+        saturating_light=1.0,
+        surface_light=1.0,
+        tidal_range=0.0,
+    ):
+        """
+        Parameters
+        ----------
+        grid: ModelGrid (RasterModelGrid, HexModelGrid, etc.)
+            A landlab grid.
+        """
+        self.extinction_coefficient = extinction_coefficient
+        self.max_carbonate_production_rate = max_carbonate_production_rate
+        self.surface_light = surface_light
+        self.tidal_range = tidal_range
+        self.saturating_light = saturating_light
+
+        super(CarbonateProducer, self).__init__(grid)
+
+        if "carbonate_production_rate" not in self.grid.at_node:
+            self.grid.add_empty("carbonate_production_rate", at="node")
+        self.grid.at_node["carbonate_production_rate"].fill(0.0)
+
+    @property
+    def extinction_coefficient(self):
+        return self._extinction_coefficient
+
+    @extinction_coefficient.setter
+    def extinction_coefficient(self, value):
+        if value >= 0.0:
+            self._extinction_coefficient = float(value)
+        else:
+            raise ValueError("must be non-negative")
+
+    @property
+    def max_carbonate_production_rate(self):
+        return self._max_carbonate_production_rate
+
+    @max_carbonate_production_rate.setter
+    def max_carbonate_production_rate(self, value):
+        if value >= 0.0:
+            self._max_carbonate_production_rate = float(value)
+        else:
+            raise ValueError("must be non-negative")
+
+    @property
+    def surface_light(self):
+        return self._surface_light
+
+    @surface_light.setter
+    def surface_light(self, value):
+        if value >= 0.0:
+            self._surface_light = float(value)
+        else:
+            raise ValueError("must be non-negative")
+
+    @property
+    def tidal_range(self):
+        return self._tidal_range
+
+    @tidal_range.setter
+    def tidal_range(self, value):
+        if value >= 0.0:
+            self._tidal_range = float(value)
+        else:
+            raise ValueError("tidal range must be non-negative")
+
+    @property
+    def saturating_light(self):
+        return self._saturating_light
+
+    @saturating_light.setter
+    def saturating_light(self, value):
+        if value >= 0.0:
+            self._saturating_light = float(value)
+        else:
+            raise ValueError("surface light must be non-negative")
+
+    @property
+    def carbonate_thickness(self):
+        return self._carb_thickness
+
+    @property
+    def sea_level(self):
+        return self.grid.at_grid["sea_level__elevation"]
+
+    @sea_level.setter
+    def sea_level(self, sea_level):
+        self.grid.at_grid["sea_level__elevation"] = sea_level
+
+    def calc_production_with_depth(self, water_depth, out=None):

```

Comment:
Great to include `out=` for economy of memory, we should make this a semi-standard practice in components!